openhcl/openhcl_boot/src/hypercall.rs

```diff
@@ -199,6 +206,90 @@ impl HvCall {
         Ok(())
     }
 
+    fn to_hv_gva_range_array(
+        memory_range: &MemoryRange,
+    ) -> ArrayVec<HvGpaRange, MAX_INPUT_ELEMENTS> {
+        const PAGES_PER_ENTRY: u64 = 2048;
+        const PAGE_SIZE: u64 = HV_PAGE_SIZE;
+        let mut ranges = ArrayVec::new();
+
+        // Calculate the total number of pages in the memory range
+        let total_pages = (memory_range.end() - memory_range.start()).div_ceil(PAGE_SIZE);
+
+        // Iterate over the memory range in chunks of 2048 pages
+        let mut current_page = memory_range.start() >> 12; // Convert start address to page number
+        let mut remaining_pages = total_pages;
+
+        while remaining_pages > 0 {
```

Comment:
I recommend doing something like `while current_page < memory_range.end_4k_gpn()` and removing `remaining_pages` as a separate variable. IMO it's almost always better to have a single loop variable.